engines/ic3sa.cpp

```diff
@@ -0,0 +1,880 @@
+/*********************                                                  */
+/*! \file ic3sa.cpp
+** \verbatim
+** Top contributors (to current version):
+**   Makai Mann
+** This file is part of the pono project.
+** Copyright (c) 2019 by the authors listed in the file AUTHORS
+** in the top-level source directory) and their institutional affiliations.
+** All rights reserved.  See the file LICENSE in the top-level source
+** directory for licensing information.\endverbatim
+**
+** \brief IC3 with Syntax-Guided Abstraction based on
+**
+**        Model Checking of Verilog RTL Using IC3 with Syntax-Guided
+**        Abstraction.
+**            -- Aman Goel, Karem Sakallah
+**
+**
+**  within Pono, we are building on the bit-level IC3 instead of directly
+**  on IC3Base, because a lot of the functionality is the same
+**  In particular, we don't need to override inductive generalization
+**
+**/
+
+#include "engines/ic3sa.h"
+
+#include "assert.h"
+#include "core/rts.h"
+#include "smt-switch/utils.h"
+#include "utils/logger.h"
+#include "utils/term_analysis.h"
+#include "utils/term_walkers.h"
+
+using namespace smt;
+using namespace std;
+
+// TODO add option to use interpolants in inductive generalization
+
+namespace pono {
+
+unordered_set<PrimOp> controllable_ops(
+    { And,
+      Or,
+      Implies,
+      // include bit-vector versions for boolector
+      // will prune out based on sort if
+      // not-applicable e.g. for other solvers
+      BVOr,
+      BVAnd });
+
+// main IC3SA implementation
+
+IC3SA::IC3SA(const Property & p,
+             const TransitionSystem & ts,
+             const smt::SmtSolver & solver,
+             PonoOptions opt)
+    : super(p, RelationalTransitionSystem(solver), solver, opt),
```

Comment:
why relational here?

Reply:
Good question. It's cause `IC3SA` learns path axioms during refinement that need to be added to the transition system, and those axioms might not be functional. I left a comment about it (way further down in the code), but I should mention it here too.

Reply:
i see later in the code that it is needed.